File: src/GnOuGo.Flow.Core/Runtime/Executors/SequenceExecutor.cs
using System.Text.Json.Nodes;
using GnOuGo.Flow.Core.Expressions;
using GnOuGo.Flow.Core.Models;

namespace GnOuGo.Flow.Core.Runtime.Executors;

/// <summary>
/// Executes steps sequentially.
/// </summary>
public sealed class SequenceExecutor : IStepExecutor
{
    public string StepType => "sequence";

    public IReadOnlyList<StepExceptionDoc>? DocumentedExceptions => new StepExceptionDoc[]
    {
        new(ErrorCodes.InputValidation, false, "The step is missing its nested `steps` collection.")
    };

    public string DslSnippet => """
        ### sequence — Execute sub-steps sequentially
        ```yaml
        - id: setup
          type: sequence
          steps:
            - id: step_a
              type: template.render
              input: { engine: mustache, template: "Hello", mode: text }
            - id: step_b
              type: template.render
              input: { engine: mustache, template: "World", mode: text }
        ```
        Output: object with each child step's output keyed by step id.
        """;

    public async Task<JsonNode?> ExecuteAsync(StepExecutionContext ctx, CancellationToken ct)
    {
        var subSteps = ctx.Step.Steps
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "sequence step requires 'steps'");

        var result = new RunResult { Success = true };
        await ctx.Engine.ExecuteStepsAsync(subSteps, ctx.Data, result, ctx.Limits, ctx.CallDepth, ctx.CallStack, ct, ctx.TelemetrySpan);

        // Return the collected step outputs
        var outputs = new JsonObject();
        foreach (var sr in result.StepResults.Where(s => s.Status == StepStatus.Succeeded))
        {
            outputs[sr.StepId] = sr.Output?.DeepClone();
        }
        return outputs;
    }
}

File: clients/python/gnougo-flow-core/src/gnougo_flow_core/runtime_steps/sequence_executor.py
from __future__ import annotations

from gnougo_flow_core.runtime import *  # noqa: F401,F403

class SequenceExecutor:
    step_type = "sequence"
    step_description = "Execute nested steps sequentially."
    dsl_snippet = """
### sequence - Execute sub-steps sequentially
```yaml
- id: setup
  type: sequence
  steps:
    - id: step_a
      type: template.render
      input:
        template: "Hello"
        mode: text
    - id: step_b
      type: template.render
      input:
        template: "World"
        mode: text
```
Output: object with each child step output keyed by step id.
"""
    documented_exceptions = [
        (ErrorCodes.INPUT_VALIDATION, False, "sequence step requires nested 'steps'."),
    ]

    async def execute_async(self, ctx: StepExecutionContext) -> Any:
        if not ctx.step.steps:
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "sequence step requires 'steps'")
        result = RunResult(success=True)
        await ctx.engine.execute_steps_async(ctx.step.steps, ctx.data, result, ctx.limits, ctx.call_depth, ctx.call_stack, ctx.telemetry_span, ct=ctx.ct)
        return {sr.step_id: sr.output for sr in result.step_results if sr.status == StepStatus.SUCCEEDED}
